baseplate/server/runtime_monitor.py

```diff
@@ -0,0 +1,142 @@
+            self.times_blocked.append(event.blocking_time)
+
+    def report(self, batch):
+        # gevent events come in on another thread. we're relying on the GIL to
+        # keep us from shenanigans here and we swap things out semi-safely to
+        # ensure minimal lost data.
+        times_blocked = self.times_blocked
+        self.times_blocked = []
+
+        for time_blocked in times_blocked:
+            batch.timer("hub_blocked").send(time_blocked)
+
+
+class _GCStatsReporter(object):
+    def report(self, batch):
+        for generation, stats in enumerate(gc.get_stats()):
+            for name, value in stats.items():
+                batch.gauge("gc.gen{}.{}".format(generation, name)).replace(value)
+
+
+class _GCTimingReporter(object):
+    def __init__(self):
+        gc.callbacks.append(self._on_gc_event)
+
+        self.gc_durations = []
+        self.current_gc_start = None
+
+    def _on_gc_event(self, phase, info):
+        if phase == "start":
+            self.current_gc_start = time.time()
+        elif phase == "stop":
+            if self.current_gc_start:
+                elapsed = time.time() - self.current_gc_start
+                self.current_gc_start = None
+                self.gc_durations.append(elapsed)
+
+    def report(self, batch):
+        gc_durations = self.gc_durations
+        self.gc_durations = []
+
+        for gc_duration in gc_durations:
+            batch.timer("gc.elapsed").send(gc_duration)
+
+
+def _report_runtime_metrics_periodically(metrics_client, reporters):
+    hostname = socket.gethostname()
+    pid = os.getpid()
+
+    while True:
+        now = time.time()
+        time_since_last_report = now % REPORT_INTERVAL_SECONDS
+        time_until_next_report = REPORT_INTERVAL_SECONDS - time_since_last_report
+        time.sleep(time_until_next_report)
+
+        try:
+            with metrics_client.batch() as batch:
+                batch.namespace += ".runtime.{}.PID{}".format(hostname, pid).encode()
+                for reporter in reporters:
+                    try:
+                        reporter.report(batch)
+                    except Exception as exc:
+                        logging.debug("Error generating server metrics: %s: %s",
+                                      reporter.__class__.__name__, exc)
+        except Exception as exc:
+            logging.debug("Error while sending server metrics: %s", exc)
+
+
+def start(server_config, application, pool):
+    metrics_client = application.baseplate._metrics_client
+    if not metrics_client:
+        logging.info("No metrics client configured. Server metrics will not be sent.")
+        return
+
+    cfg = config.parse_config(server_config, {
+        "monitoring": {
+            "blocked_hub": config.Optional(config.Timespan, default=None),
+            "concurrency": config.Optional(config.Boolean, default=True),
+            "gc": {
+                "stats": config.Optional(config.Boolean, default=True),
```

Comment:
Should this default to `False` for python 2? Or even better... should we remove this option from the config entirely if we are running python 2?

Reply:
I believe that you've mentioned removing python 2 in the next release. I guess this doesn't matter if that's true

Reply:
Yeah, there's some issues with this on older versions of gevent and python. I'm going to clean that up and add tests.